### PR TITLE
Landing optimizations

### LINKED ISSUE #4657: Perf optimization on landing page
### Is your feature request related to a problem?

When I open the landing page and have not scroll down to see more content, I could see lots of request were sent and one of them take severals seconds. A lots of image were requested at the same time. Both of them would increase the response time.        

### Describe the solution you would like

1. merge some of graphql query request into one graphql
2. Avoid big query: write a resolver indexer side would help
3. cache: how about caching the data for Spotlight and top collectioin on localstorage.
4. lazy load image

### Describe alternatives you have considered

-

### Screenshots

image

image

### Are there opened related issues?

https://github.com/kodadot/nft-gallery/issues/2816

### PR DESCRIPTION
**Pull request by shashkovdanil**
## PR Type

Refactoring, Performance optimizations

## Context

Closes #4657

#### Had issue bounty label?

- [ ] Fill up your KSM address: [Payout](https://beta.kodadot.xyz/transfer?target=bXn3ExwLs8rpcpknp9wNqnNVgEYVLnTtFPDrJtaPHYds7Vfjv&usdamount=0&donation=true>)

#### What I've done

1. Optimized fonts using `@nuxtjs/google-fonts`

Before (360kb):
font-before
After (50kb):
font-after

2. Optimized image using `@nuxt/image` and converting `png` to `webp`, for example

Before (650kb):
blurred-before
After (3kb):
blurred-after

3. Huge [`highlight.js`](https://bundlephobia.com/package/highlight.js@11.9.0) to lightweight [`Prism`](https://bundlephobia.com/package/prismjs@1.29.0)

4. Load huge `@google/model-viewer` only if the `ModelMedia` component is in the viewport

5. Fixed lodash imports (only `import someFn from 'lodash/someFn'`)

6. Disabled prefetch for nuxt links. It's really big nuxt js problem: by default nuxt does prefetch for all pages and layouts, which is very bad for performance as we load a ton of unused JavaScript, currently there is no solution other than disabling prefetch for links manually. This problem is faced by many nuxt users, here is the issue https://github.com/nuxt/nuxt/issues/13778

### CODE REVIEW

**Comment by Gavin-Gong** [comment]
Looks like the page has lost some CSS styles.

image

**Comment by yangwao** [comment]
related?
- https://github.com/kodadot/nft-gallery/pull/7924

**Comment by shashkovdanil** [comment]
@Gavin-Gong please tell me how to reproduce it, or maybe you have errors in the console?
Screenshot 2023-11-13 at 11 01 31

**Comment by shashkovdanil** [comment]
> Related as noticed adding `@nuxt/image` ?
> 
> * [add @nuxt/image #7924](https://github.com/kodadot/nft-gallery/pull/7924)

@yangwao ye it's related, I think we can keep my version, I've tested it in different browsers, also configured caching

**Comment by yangwao** [comment]
> or maybe you have errors in the console?

current preview looks like this
- https://deploy-preview-8030--koda-canary.netlify.app/

image

**Comment by preschian** [comment]
tbh, I would prefer to revisit the issue after SSR is supported

**Review by preschian — CHANGES_REQUESTED** [review]
> tbh, I would prefer to revisit the issue after SSR is supported


up

**Comment by shashkovdanil** [comment]
@preschian I can help with SSR task if you don't mind. I've had a lot of experience with it

**Comment by yangwao** [comment]
> I can help with SSR task

here was iirc some work done 
- https://github.com/kodadot/nft-gallery/pull/7725

**Comment by preschian** [comment]
@shashkovdanil you can keep these improvements if you want:
```
1. Optimized fonts using @nuxtjs/google-fonts
3. Huge highlight.js to lightweight Prism
4. Load huge @google/model-viewer only if the ModelMedia component is in the viewport
5. Fixed lodash imports (only import someFn from 'lodash/someFn')
```

but please, skip prefetch stuff for now

File: libs/ui/src/components/MediaItem/type/ImageMedia.vue [hunk 1]
```diff
@@ -5,11 +5,15 @@
       'is-square image': !original,
       'is-detail': isDetail,
     }">
-    <img
+    <NuxtImg
       :src="src"
       class="is-block image-media__image no-border-radius"
       :alt="alt"
       data-testid="type-image"
+      loading="lazy"
```
Comment (yangwao) [inline libs/ui/src/components/MediaItem/type/ImageMedia.vue:13]: Missing provider ? Like in #7924  
- https://github.com/kodadot/nft-gallery/pull/7924/files#diff-01c20791d8b8c84879140e2fc9dd819eb73f19d5fc62aff1053387e6658f4607R10

**Review by yangwao — APPROVED** [review]
I've noticed that #7924 loads on explorer items much faster than this PR, curious why

File: libs/ui/src/components/MediaItem/type/ImageMedia.vue (hunk 1 above)
Comment (shashkovdanil) [reply libs/ui/src/components/MediaItem/type/ImageMedia.vue:13]: https://github.com/kodadot/nft-gallery/pull/8030/files#diff-5977891bf10802cdd3cde62f0355105a1662e65b02ae4fb404a27bb0f5f53a07R255

**Comment by shashkovdanil** [comment]
> @shashkovdanil you can keep these improvements if you want:
> 
> ```
> 1. Optimized fonts using @nuxtjs/google-fonts
> 3. Huge highlight.js to lightweight Prism
> 4. Load huge @google/model-viewer only if the ModelMedia component is in the viewport
> 5. Fixed lodash imports (only import someFn from 'lodash/someFn')
> ```
> 
> but please, skip prefetch stuff for now

@preschian okay, ye, removing prefetch is risky, so I did it in a separate commit, okay, I'll revert it

File: libs/ui/src/components/MediaItem/type/ImageMedia.vue (hunk 1 above)
Comment (yangwao) [reply libs/ui/src/components/MediaItem/type/ImageMedia.vue:13]: I think we have Cloudflare, not IPX
- https://image.nuxt.com/providers/cloudflare
- https://image.nuxt.com/providers/ipx

**Comment by shashkovdanil** [comment]
> @shashkovdanil you can keep these improvements if you want:
> 
> ```
> 1. Optimized fonts using @nuxtjs/google-fonts
> 3. Huge highlight.js to lightweight Prism
> 4. Load huge @google/model-viewer only if the ModelMedia component is in the viewport
> 5. Fixed lodash imports (only import someFn from 'lodash/someFn')
> ```
> 
> but please, skip prefetch stuff for now

Done!

File: libs/ui/src/components/MediaItem/type/ImageMedia.vue (hunk 1 above)
Comment (shashkovdanil) [reply libs/ui/src/components/MediaItem/type/ImageMedia.vue:13]: Removed nuxt-image from this PR

File: nuxt.config.ts [hunk 2]
```diff
@@ -225,8 +222,28 @@ export default defineNuxtConfig({
     '@pinia/nuxt',
     '@nuxt/content',
     'nuxt-simple-sitemap',
+    '@nuxtjs/google-fonts',
+    'nuxt-purgecss',
```
Comment (preschian) [inline nuxt.config.ts:226]: still broken on deployment preview, I suspect because of this module. let's try it by removing the `nuxt-purgecss` module

Comment (shashkovdanil) [reply nuxt.config.ts]: ye, it was because of this, it has `enable: false` for dev env, so I didn't notice problem

**Review by preschian — APPROVED** [review]
thank you for the improvements 🚀

File: libs/ui/src/components/MediaItem/type/ImageMedia.vue (hunk 1 above)
Comment (yangwao) [reply libs/ui/src/components/MediaItem/type/ImageMedia.vue:13]: > Removed nuxt-image from this PR

yeah atomic small PR preferable, feel free do it another PR if so

**Comment by yangwao** [comment]
We've lost some accessibility, best practices and SEO points apparently in this PR 👀

canary.kodadot.xyz
image

this PR
image

**Comment by shashkovdanil** [comment]
> We've lost some accessibility, best practices and SEO points apparently in this PR 👀
> 
> canary.kodadot.xyz image
> 
> this PR image

@yangwao It won't affect SEO in any way. The thing is that before the images for Article were made through div with background-image, I changed them to img with loading="lazy", but did not add alt attribute, because of this the ranking decreased. Now I will add alt

**Comment by yangwao** [comment]
> Now I will add alt

image
works!

**Comment by shashkovdanil** [comment]
@prury @preschian I need advice, search.spec crashes because there is this code Array.from(document.images).every((i) => i.complete), which checks if ALL images from the page are loaded. But since loading="lazy" was added to improve performance, only those images that are in the viewport are loaded.

What's the best way to fix the test? Why does the test wait for all images to load? Can this be removed?

**Comment by prury** [comment]
> @prury @preschian I need advice, search.spec crashes because there is this code `Array.from(document.images).every((i) => i.complete)`, which checks if ALL images from the page are loaded. But since loading="lazy" was added to improve performance, only those images that are in the viewport are loaded.
> 
> What's the best way to fix the test? Why does the test wait for all images to load? Can this be removed?

from what i could verify, if you go to RMRK2, search for waifu, it only provides search results for collections and not NFTs

**Comment by preschian** [comment]
> But since loading="lazy"

I prefer to remove lazy on `libs/ui/src/components/MediaItem/type/ImageMedia.vue` and `libs/ui/src/components/MediaItem/type/ModelMedia.vue`. Because the component can be rendered on above the fold section https://web.dev/articles/lazy-loading-best-practices#:~:text=Anything%20resting%20above%20the%20fold,finished%20loading%20and%20begin%20execution

We can put `loading="lazy"` if we are sure that the component is below the fold

**Comment by shashkovdanil** [comment]
Ok, I removed loading="lazy", we can think later how to implement it more intelligently

File: libs/ui/src/components/MediaItem/MediaItem.vue [hunk 3]
```diff
@@ -92,6 +91,35 @@ const props = withDefaults(
...
+let isModelVisible = ref(false)
+let observer: IntersectionObserver
+
+onMounted(() => {
+  observer = new IntersectionObserver((entries) => {
+    if (entries.some((entry) => entry.isIntersecting)) {
+      if (mimeType.value === 'model/gltf-binary') {
+        isModelVisible.value = true
+        modelComponent = defineAsyncComponent(
+          () => import('./type/ModelMedia.vue'),
```
Comment (preschian) [inline libs/ui/src/components/MediaItem/MediaItem.vue:105]: need to make sure this one, sometimes the components were not loaded

Screenshot 2023-11-13 at 10 40 14 PM

https://deploy-preview-8030--koda-canary.netlify.app/ksm/explore/items?listed=true

Comment (shashkovdanil) [reply libs/ui/src/components/MediaItem/MediaItem.vue]: Done

**Comment by shashkovdanil** [comment]
@preschian jfyi it's ready for merge

**Comment by prury** [comment]
Talisman wallet image won't load the first time you try to connect wallet on desktop: 

https://github.com/kodadot/nft-gallery/assets/36627808/500e8835-d97c-4156-a15a-be78254894fa

**Comment by yangwao** [comment]
Seems fonts has changed as @prury reported 

canary
image
pr
image

**Comment by prury** [comment]
@shashkovdanil sorry for the trouble, but connect wallet problem persists

**Comment by shashkovdanil** [comment]
@prury ufff it was crazy bug. If you import the `chain` function as `import { chain } from 'lodash'`, everything works fine. But then we're downloading 200kb of lodash. If we import this function correctly, using `import chain from 'lodash/chain'`, then on the first call we get the error that `chain(...).map is not function`. I just rewrote the code without using the chain function

**Comment by prury** [comment]
> @prury ufff it was crazy bug. If you import the `chain` function as `import { chain } from 'lodash'`, everything works fine. But then we're downloading 200kb of lodash. If we import this function correctly, using `import chain from 'lodash/chain'`, then on the first call we get the error that `chain(...).map is not function`. I just rewrote the code without using the chain function

oh 😄, thank you for rewriting, wallet bug is now gone

**Comment by yangwao** [comment]
Great, we've gained in performance, yet it seems we've lost some points in best practices, you might have a look @shashkovdanil, please ?

image

**Review by Jarsen136 — APPROVED** [review]
code lgtm

**Comment by yangwao** [comment]
> Fix keyboard navigation, put a11y attributes everywhere. If you have a request for this

I think mostly performance would help keep it on green currently LCP, CLS and so on
- https://github.com/kodadot/nft-gallery/issues/6322

**Comment by yangwao** [comment]
Did not notice anything broken, let's see, merging it, great PR @shashkovdanil thanks!

pay 80 usd

**Comment by yangwao** [comment]
pay 80 usd

updated your address based on previous PR

**Comment by yangwao** [comment]
😍 Perfect, I’ve sent the payout 
💵 $80 @ 5.44 USD/DOT ~ 14.706 $DOT
🧗 [16SpvdDgKNiQ3c53DxX7refnQcKUD3uRNim3Z1HBJLNGrtSP](https://polkadot.subscan.io/account/16SpvdDgKNiQ3c53DxX7refnQcKUD3uRNim3Z1HBJLNGrtSP)
🔗 [0xded7066fa6caa4986584a1dc2ed7cf9703fe27c2efd591b906ba137693a4e312](https://polkadot.subscan.io/extrinsic/0xded7066fa6caa4986584a1dc2ed7cf9703fe27c2efd591b906ba137693a4e312)

🪅 Let’s grab another issue and get rewarded! 
🪄 [github.com/kodadot/nft-gallery/issues](https://github.com/kodadot/nft-gallery/issues/)

**Pull request merged**